lib/basedriver/commands/find.js

```diff
@@ -357,42 +367,119 @@ helpers.getScreenshotForImageFind = async function (screenWidth, screenHeight) {
   // of coordinates returned by the image match algorithm, since we match based
   // on the screenshot coordinates not the device coordinates themselves. There
   // are two potential types of mismatch: aspect ratio mismatch and scale
-  // mismatch.
+  // mismatch. We need to detect and fix both
+
+  const scale = {xScale: 1.0, yScale: 1.0};
+
+  const screenAR = screenWidth / screenHeight;
+  const shotAR = shotWidth / shotHeight;
+  if (Math.round(screenAR * FLOAT_PRECISION) === Math.round(shotAR * FLOAT_PRECISION)) {
+    log.info('Screenshot aspect ratio matched screen aspect ratio');
+  } else {
+    log.warn(`When trying to find an element, determined that the screen ` +
+             `aspect ratio and screenshot aspect ratio are different. Screen ` +
+             `is ${screenWidth}x${screenHeight} whereas screenshot is ` +
+             `${shotWidth}x${shotHeight}.`);
+
+    // Select smaller ratio
+    // this.getScreenshot(shotWidth, shotHeight) is 540x397,
+    // this.getDeviceSize(screenWidth, screenHeight) is 1080x1920.
+    // The ratio is {xScale: 0.5, yScale: 0.2}.
+    // In this case, we must choose `yScale: 0.2` as scaleFactor.
+    // Because if Appium selects the both ratio, the screenshot will be distorted.
+    // If Appium selects the xScale, height will be bigger than real screenshot size
+    // which is used to image comparison by OpenCV as a base image.
+    const xScale = (1.0 * shotWidth) / screenWidth;
+    const yScale = (1.0 * shotHeight) / screenHeight;
+    const scaleFactor = xScale >= yScale ? yScale : xScale;
+
+    log.warn(`Resizing screenshot to ${shotWidth * scaleFactor}x${shotHeight * scaleFactor} to match ` +
+             `screen aspect ratio so that image element coordinates have a ` +
+             `greater chance of being correct.`);
+    imgObj = imgObj.resize(shotWidth * scaleFactor, shotHeight * scaleFactor);
+
+    scale.xScale *= scaleFactor;
+    scale.yScale *= scaleFactor;
+  }
 
   // Resize based on the screen dimensions only if both width and height are mismatched
   // since except for that, it might be a situation which is different window rect and
   // screenshot size like `@driver.window_rect #=>x=0, y=0, width=1080, height=1794` and
   // `"deviceScreenSize"=>"1080x1920"`
-  let scale;
   if (screenWidth !== shotWidth && screenHeight !== shotHeight) {
```

Comment:
shouldn't we be comparing the _new_ imgObj width and height here? it looks like we're comparing against the old shotWidth/shotHeight, but those could have changed in the previous block, right?

Reply:
Right. I added https://github.com/appium/appium-base-driver/pull/307/commits/b536b1039766f45e95ee5be3c79d68a3f9e84304 to update `shotWidth` and `shotHeight` if they should be updated in the ratio comparison

Reply:
@jlipps Do you have a chance to take a look ^?